amgcl/detail/qr_solve.hpp

```diff
@@ -0,0 +1,78 @@
+#ifndef AMGCL_DETAIL_QR_SOLVE_HPP_
+#define AMGCL_DETAIL_QR_SOLVE_HPP_
+
+#include <vector>
+#include <cmath>
+
+#include <amgcl/util.hpp>
+#include <amgcl/value_type/interface.hpp>
+#include <amgcl/detail/qr.hpp>
+
+namespace amgcl {
+namespace detail {
+
+enum matrix_shape {
+	tall_matrix,
+	wide_matrix
+};
+
+/// QR solver.
+template <typename value_type, storage_order Order>
+class QRSolve {
+protected:
+	typedef typename amgcl::math::rhs_of<value_type>::type rhs_type;
+
+	int m, n, k, nmax;
+	matrix_shape shape;
+	QR<value_type, Order> qr;
+
+public:
+
+	QRSolve(int rows, int cols, value_type *A, int max_cols = -1) {
+		m = rows;
+		n = cols;
+		k = std::min(m, n);
+		nmax = (max_cols < 0 ? cols : max_cols);
+		shape = m >= n ? tall_matrix : wide_matrix;
+
+		if (shape == tall_matrix)
+			qr.compute(m, n, A, max_cols);
+		else
+			qr.compute(n, m, transpose(A));
+	}
+
+	// Transposes a matrix stored linearly.
+	value_type* transpose(value_type *A) {
+		const int row_stride = (Order == row_major ? nmax : 1);
+		const int col_stride = (Order == row_major ? 1 : m);
+
+		const int row_stride_t = (Order == row_major ? m : 1);
+		const int col_stride_t = (Order == row_major ? 1 : nmax);
+
+		value_type* AT = new value_type[m*n];
```

Comment:
This is a memory leak, since AT is never deleted. Also, I don't think we need to do the transposition explicitly, as QR methods operate with row and col strides. To elaborate, if a matrix `A[n x m]` is stored in row-major order, we can access `A(i,j)` by doing `A.data[i * row_stride + j * col_stride]`. Now if we want to transpose the matrix, we can leave its `data` array intact, and just swap its `row_stride` with `col_stride`.